grails-app/services/org/pih/warehouse/importer/InventoryLevelImportDataService.groovy

```diff
@@ -10,27 +10,217 @@
+        if (!product) {
+            throw new IllegalArgumentException("Product with product code ${params.productCode} was not found")
+        }
+
+        // Validate that the facility exists
+        Location facility = Location.findByName(params.inventory as String)
+        if (!facility || !facility.inventory) {
+            throw new IllegalArgumentException("Facility with name ${params.inventory} was not found or is not valid")
+        }
+
+        // TODO Any record with a non-empty internal location is a bin replenishment rule rather than an facility-based
+        //  inventory level rule. In order to protect the bin replenishment rules from being overwritten with this
+        //  data import feature, we're going to skip these records for now. The best way to deal with these will be to
+        //  create a separate importer/exporter to handle the bin replenishment rules separately. Or we can remove
+        //  this check and ensure that the following code handles these rules properly (i.e. most of the time the
+        //  bin replenishment only needs internal location, min quantity, reorder quantity, max quantity). Therefore,
+        //  we should create separate command classes to validate the two types of inventory levels.
+        if (params.internalLocation) {
+            log.info "Skipping update to bin replenishment rule ${params}"
+            return
+        }
+
+        // Resolve any objects that require a database lookup
+        params.internalLocation = params.internalLocation ? resolveInternalLocation(facility, params.internalLocation as String) : null
+        params.preferredBinLocation = params.preferredBinLocation ? resolveInternalLocation(facility, params.preferredBinLocation as String) : null
+        params.replenishmentLocation = params.replenishmentLocation ? resolveInternalLocation(facility, params.replenishmentLocation as String) : null
+
+        // Update the inventory level
+        InventoryLevel inventoryLevel = updateOrCreateInventoryLevel(facility, product, params)
+
+        // Delete the inventory level if the status has been set to inactive,
+        if (inventoryLevel.status == InventoryStatus.INACTIVE) {
+            product.removeFromInventoryLevels(inventoryLevel)
+            inventoryLevel.delete()
+            product.save()
+        } else {
+            // Save inventory level OR we could return the inventory level and save them all at the end
+            product.addToInventoryLevels(inventoryLevel)
+            inventoryLevel.save()
+            product.save()
+        }
+    }
+
+    Location resolveInternalLocation(Location facility, String internalLocationName) {
+        Location internalLocation = facility.getInternalLocation(internalLocationName)
+        if (!internalLocation) {
+            throw new IllegalArgumentException("Unable to locate internal location ${internalLocationName} within facility ${facility?.name}")
+        }
+        return internalLocation
```

Comment:
Can't we handle the lack of `params.internalLocation` here instead of checking it three times when assigning to:
```
params.internalLocation = params.internalLocation ? resolveInternalLocation(facility, params.internalLocation as String) : null
params.preferredBinLocation = params.preferredBinLocation ? resolveInternalLocation(facility, params.preferredBinLocation as String) : null
params.replenishmentLocation = params.replenishmentLocation ? resolveInternalLocation(facility, params.replenishmentLocation as String) : null
```

Reply:
Can you provide an example for what you mean? Something like this?
```
params.internalLocation = resolveInternalLocation(facility, params.internalLocation as String)
```

Reply:
I meant adding 
```
if (internalLocationName == null) {
   return null
}
```
at the beginning of the resolveInternalLocation method, so you don't have to check it while assigning to params.XYZ

Reply:
Yeah that's what i was thinking you meant. It feels weird to deal with that in the resolve method since it seems to be the API client's responsibility to decide how it wants to deal with that case. A better approach might be to pass a "default" value to the resolve method, so I'll consider doing that.